src/main/java/jcifs/dcerpc/msrpc/lsarpc.java

```diff
@@ -0,0 +1,1046 @@
+                    this.name.buffer = new short[_name_buffers];
+                }
+                _src = _src.derive(_name_bufferi);
+                for ( int _i = 0; _i < _name_bufferl; _i++ ) {
+                    this.name.buffer[ _i ] = (short) _src.dec_ndr_short();
+                }
+            }
+            if ( _dns_domain_bufferp != 0 ) {
+                _src = _src.deferred;
+                int _dns_domain_buffers = _src.dec_ndr_long();
+                _src.dec_ndr_long();
+                int _dns_domain_bufferl = _src.dec_ndr_long();
+                int _dns_domain_bufferi = _src.index;
+                _src.advance(2 * _dns_domain_bufferl);
+
+                if ( this.dns_domain.buffer == null ) {
+                    if ( _dns_domain_buffers < 0 || _dns_domain_buffers > 0xFFFF )
+                        throw new NdrException(NdrException.INVALID_CONFORMANCE);
+                    this.dns_domain.buffer = new short[_dns_domain_buffers];
+                }
+                _src = _src.derive(_dns_domain_bufferi);
+                for ( int _i = 0; _i < _dns_domain_bufferl; _i++ ) {
+                    this.dns_domain.buffer[ _i ] = (short) _src.dec_ndr_short();
+                }
+            }
+            if ( _dns_forest_bufferp != 0 ) {
+                _src = _src.deferred;
+                int _dns_forest_buffers = _src.dec_ndr_long();
+                _src.dec_ndr_long();
+                int _dns_forest_bufferl = _src.dec_ndr_long();
+                int _dns_forest_bufferi = _src.index;
+                _src.advance(2 * _dns_forest_bufferl);
+
+                if ( this.dns_forest.buffer == null ) {
+                    if ( _dns_forest_buffers < 0 || _dns_forest_buffers > 0xFFFF )
+                        throw new NdrException(NdrException.INVALID_CONFORMANCE);
+                    this.dns_forest.buffer = new short[_dns_forest_buffers];
+                }
+                _src = _src.derive(_dns_forest_bufferi);
+                for ( int _i = 0; _i < _dns_forest_bufferl; _i++ ) {
+                    this.dns_forest.buffer[ _i ] = (short) _src.dec_ndr_short();
+                }
+            }
+            if ( this.domain_guid.node == null ) {
+                if ( _domain_guid_nodes < 0 || _domain_guid_nodes > 0xFFFF )
+                    throw new NdrException(NdrException.INVALID_CONFORMANCE);
+                this.domain_guid.node = new byte[_domain_guid_nodes];
+            }
+            _src = _src.derive(_domain_guid_nodei);
+            for ( int _i = 0; _i < _domain_guid_nodes; _i++ ) {
+                this.domain_guid.node[ _i ] = (byte) _src.dec_ndr_small();
+            }
+            if ( _sidp != 0 ) {
+                if ( this.sid == null ) { /* YOYOYO */
+                    this.sid = new rpc.sid_t();
+                }
+                _src = _src.deferred;
+                this.sid.decode(_src);
+
+            }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 388:395, 429:436.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.